ownCloud/Client/Actions/Actions+Extensions/UploadPhotosAction.swift

```diff
@@ -95,59 +135,109 @@ class UploadPhotosAction: UploadBaseAction {
 
-		let ressources = PHAssetResource.assetResources(for: asset)
+		guard let rootItem = context.items.first else { return }
 
-		if let ressource = ressources.first, let rootItem = context.items.first {
-			let filename = ressource.originalFilename
+		// Prepare progress object for importing full size asset from photo library
+		let progress = Progress(totalUnitCount: 100)
+		progress.localizedDescription = "Importing from photo library".localized
+		self.publish(progress: progress)
 
-			let progress = Progress(totalUnitCount: 100)
-			progress.localizedDescription = String(format: "Importing '%@' from photo library".localized, filename)
+		// Setup import options, allow download asset from network if necessary
+		let contentInputOptions = PHContentEditingInputRequestOptions()
+		contentInputOptions.isNetworkAccessAllowed = true
+		contentInputOptions.progressHandler = { (percentage:Double, _) in
+			progress.completedUnitCount = Int64(percentage * 100)
+		}
 
-			let options = PHAssetResourceRequestOptions()
-			options.isNetworkAccessAllowed = true
-			options.progressHandler = { (completed:Double) in
-				progress.completedUnitCount = Int64(completed * 100)
-			}
+		// Import full size asset
+		asset.requestContentEditingInput(with: contentInputOptions) { (input, _) in
+			self.unpublish(progress: progress)
 
-			self.publish(progress: progress)
+			self.uploadSerialQueue.async {
+				if let input = input, let url = input.fullSizeImageURL {
 
-			var localURL = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(filename)
+					func performUpload(sourceURL:URL, copySource:Bool) {
 
-			let contentInputOptions = PHContentEditingInputRequestOptions()
-			contentInputOptions.isNetworkAccessAllowed = true
-			asset.requestContentEditingInput(with: contentInputOptions) { (input, _) in
-				if let fullImage = CIImage(contentsOf: input!.fullSizeImageURL!) {
-					let storeAsHEIF = ressource.uniformTypeIdentifier == "public.heic" && !userDefaults.convertHeic
-					let colorSpace = CGColorSpaceCreateDeviceRGB()
+						@discardableResult func removeSourceFile() -> Bool {
+							do {
+								try FileManager.default.removeItem(at: sourceURL)
+								return true
+							} catch {
+								return false
+							}
+						}
 
-					var imageData : Data?
-					if storeAsHEIF {
-						imageData = CIContext().heifRepresentation(of: fullImage, format: CIFormat.RGBA8, colorSpace: colorSpace)
-						print(fullImage.properties)
-					} else {
-						localURL = localURL.deletingPathExtension().appendingPathExtension("jpg")
-						imageData = CIContext().jpegRepresentation(of: fullImage, colorSpace: colorSpace)
+						let fileName = sourceURL.lastPathComponent
+						self.upload(itemURL: sourceURL, to: rootItem, name:fileName, completionHandler: { ( success, _) in
+							completion(success, true)
+						}, placeholderHandler: { (_, error) in
+							if copySource && error != nil {
```

Comment:
I believe this should be `!copySource` instead
```suggestion
							if !copySource && error != nil {
```

Reply:
@felix-schwarz yes, right. Fixed.